### PR TITLE
lapack: add Dlag2 subroutine

### PR DESCRIPTION
**Pull request by soypat**
Please take a look. Process of solving #1651.

Dlag2 solves a 2x2 eigenvalue problem of the form `|A-xB| = 0` where `B` is upper triangular.

### CODE REVIEW

File: lapack/gonum/dlag2.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+		panic(shortA)
+	case len(b) < ldb+2:
+		panic(shortB)
+	}
+
+	const (
+		safmin = dlamchS
+		fuzzy1 = 1 + 1e-5
+	)
+	rtmin := math.Sqrt(safmin)
```
Comment (vladimir-ch) [inline lapack/gonum/dlag2.go:77]: It's unfortunate that our `drtmin` is slightly different.

File: lapack/gonum/dlag2.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+		wbig := shift + sum
+
+		// Compute smaller eigenvalue.
+		var wsmall, wdet float64
+		wsmall = shift + diff
+		if 0.5*math.Abs(wbig) > math.Max(math.Abs(wsmall), safmin) {
+			wdet = (a11*a22 - a12*a21) * (binv11 * binv22)
+			wsmall = wdet / wbig
+		}
+		// Choose (real) eigenvalue closest to 2,2 element of A*B**(-1) for WR1.
```
Comment (vladimir-ch) [inline lapack/gonum/dlag2.go:166]:
```suggestion
		// Choose (real) eigenvalue closest to 2,2 element of A*B^{-1} for wr1.
```

File: lapack/testlapack/dlag2.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"fmt"
+	"math"
+	"testing"
+
+	"golang.org/x/exp/rand"
+)
+
+type Dlag2er interface {
+	// DLAG2 computes the eigenvalues of a 2 x 2 generalized eigenvalue
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:16]: Please remove this comment, there's no point in documenting these helper interfaces.

File: lapack/testlapack/dlag2.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	// than that, then  +/- sqrt(SAFMIN) will be used instead of
+	// that diagonal.
+	//
+	// Dlag2 is an internal routine. It is exported for testing purposes.
+	Dlag2(a []float64, lda int, b []float64, ldb int) (scale1, scale2, wr1, wr2, wi float64)
+}
+
+func Dlag2Test(t *testing.T, impl Dlag2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for _, lda := range []int{2, 3, 4, 5, 6, 7, 30} {
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:39]: There's no benefit from testing so many strides, 2 and say 5 is enough. On the other hand, you can add a for loop that runs for example 10-100 times to get some variability in the random test data.

File: lapack/testlapack/dlag2.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			dlag2Test(t, impl, rnd, lda, ldb)
+		}
+	}
+}
+
+func dlag2Test(t *testing.T, impl Dlag2er, rnd *rand.Rand, lda, ldb int) {
+	const tol = 1e-14
+	a := randomGeneral(2, 2, lda, rnd)
+	b := randomGeneral(2, 2, ldb, rnd)
+	b.Data[ldb] = 0 // Dlag2 requires b to be upper triangular
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:50]: There is no requirement in the code that this element is set to zero, it's simply not referenced by Dlag2. The test will be in fact better if you leave it set to a random value in case our implementation does reference it by mistake.

File: lapack/testlapack/dlag2.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	a11 := a.Data[0]
+	a12 := a.Data[1]
+	a21 := a.Data[lda]
+	a22 := a.Data[lda+1]
+	b11 := b.Data[0]
+	b12 := b.Data[1]
+	b22 := b.Data[ldb+1]
+	name := fmt.Sprintf("a11=%.4v,a12=%.4v,a21=%.4v,a22=%.4v,b11=%.4v,b12=%.4v,b22=%.4v", a11, a12, a21, a22, b11, b12, b22)
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:58]: I'm not sure if this is useful. If there's an error, having the elements printed to 4 decimal places will not help.

File: lapack/testlapack/dlag2.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	for _, lda := range []int{2, 3, 4, 5, 6, 7, 30} {
+		for _, ldb := range []int{2, 3, 4, 5, 6, 7, 30} {
+			dlag2Test(t, impl, rnd, lda, ldb)
+		}
+	}
+}
+
+func dlag2Test(t *testing.T, impl Dlag2er, rnd *rand.Rand, lda, ldb int) {
+	const tol = 1e-14
+	a := randomGeneral(2, 2, lda, rnd)
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:48]: Generating A and B randomly is ok but I'm afraid it will not cover those parts of the code that handle under/overflow. Have you looked what the coverage of Dlag2 is which this test?

You should take a look at `TESTING/EIG/dget53.f` and how they check the eigenvalues. See also where in `TESTING/EIG` this function is being called and how the input is being generated. Maybe you can find how to generate a difficult input to Dlag2.

File: lapack/testlapack/dlag2.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		t.Errorf("%v: got a residual from |%0.4v*A - w1*B| > tol: %v", name, scale1, res1)
+	}
+	res2 := det2x2(scale2*a11-wr2*b11, scale2*a12-wr2*b12,
+		scale2*a21, scale2*a22-wr2*b22)
+	if math.Abs(res2) > tol {
+		t.Errorf("%v: got a residual from |s*A - w2*B| > tol: %v", name, res1)
+	}
+}
+
+// Solves complex determinant of
```
Comment (vladimir-ch) [inline lapack/testlapack/dlag2.go:93]:
```suggestion
// cmplxdet2x2 returns the determinant of
```
and please the body of the function on its own line.

File: lapack/gonum/dlag2.go (hunk 1 above)
Comment (soypat) [reply lapack/gonum/dlag2.go:77]: Should I change this to be drtmin?

Comment (vladimir-ch) [reply lapack/gonum/dlag2.go:79]: No, it was just a comment.

File: lapack/testlapack/dlag2.go (hunk 7 above)
Comment (soypat) [reply lapack/testlapack/dlag2.go:33]: I've added `dget53Test` to the test but I'm struggling reading the tests. Could we add these edge case tests to the backlog?

Comment (vladimir-ch) [reply lapack/testlapack/dlag2.go:33]: Yeah, no worries. I'll merge what you have here now and then take a look myself and probably submit a clean-up PR later.

**Review by vladimir-ch — APPROVED** [review]
Thanks.
